### PR TITLE
Restore dealias Nil in Cleanup

### PR DESCRIPTION
**Pull request by som-snytt**
Dealiasing Nil, List, Seq was previously in Typer.

Test fixes for jvm18 which disallows setting security manager.

Null check check checks aliased module.

Change in reflection for jvm18 requires .check pragma.

### CODE REVIEW

File: test/junit/scala/tools/nsc/backend/jvm/BytecodeTest.scala [hunk 1]
```diff
@@ -529,6 +529,41 @@ class BytecodeTest extends BytecodeTesting {
...
+Jump(GOTO, Label(34)),
+Label(34),
+TypeOp(NEW, "scala/MatchError"),
+Op(DUP),
+VarOp(ALOAD, 3),
+Invoke(INVOKESPECIAL, "scala/MatchError", "<init>", "(Ljava/lang/Object;)V", false),
+Op(ATHROW),
+
+
+
```
Comment (som-snytt) [inline test/junit/scala/tools/nsc/backend/jvm/BytecodeTest.scala]: To be reformatted. The test diff output makes updating the test difficult. Oh, I guess it's called diff for difficult.

**Comment by som-snytt** [comment]
Oh noes!
```
[error] java.io.IOException: No space left on device
```

**Comment by som-snytt** [comment]
- [ ] Fix tests for small List/Seq optimization
- [x] Fix small List/Seq optimization

Needed to strip `asInstanceOf` from the apply. Periodic reminder that it's intended to work only for reference types!

**Comment by som-snytt** [comment]
This includes the run/t6989.check for JVM18 that I need right now.

**Comment by SethTisue** [comment]
@lrytz you reviewed #9969, want to grab this one too?

**Review by lrytz — APPROVED** [review]
Thank you!

**Comment by som-snytt** [comment]
I'll follow up with what the allocation tests were actually testing, and with bytecode tests that Lukas asked for on a related PR.

**Pull request merged**